src/python_be.cc

```diff
@@ -1961,64 +1319,7 @@ ModelInstanceState::ProcessRequests(
 
 ModelInstanceState::~ModelInstanceState()
 {
-  if (initialized_) {
-    {
-      bi::scoped_lock<bi::interprocess_mutex> lock(*health_mutex_);
-      ipc_control_->stub_health = false;
-    }
-
-    // Sleep 1 second so that the child process has a chance to change the
-    // health variable
-    sleep(1);
-
-    bool healthy = false;
-    bool force_kill = false;
-    {
-      bi::scoped_lock<bi::interprocess_mutex> lock(*health_mutex_);
-      healthy = ipc_control_->stub_health;
-    }
-
-    if (healthy) {
-      // Finalize command does not have any arguments.
-      std::unique_ptr<IPCMessage> ipc_message =
-          IPCMessage::Create(shm_pool_, false /* inline_response */);
-
-      ModelState* model_state = reinterpret_cast<ModelState*>(Model());
-      if (model_state->IsDecoupled()) {
-        futures_.clear();
-        parent_message_queue_->Push(DUMMY_MESSAGE);
-        decoupled_monitor_.join();
-      }
-
-      // Wait for all the futures to be finished.
-      thread_pool_->wait();
-
-      ipc_message->Command() = PYTHONSTUB_FinalizeRequest;
-      stub_message_queue_->Push(ipc_message->ShmHandle());
-      parent_message_queue_->Pop();
-
-
-      stub_message_queue_.reset();
-      parent_message_queue_.reset();
-      memory_manager_.reset();
-
-    } else {
-      force_kill = true;
-    }
-
-    int status;
-    if (force_kill) {
-      kill(stub_pid_, SIGKILL);
-    }
-    waitpid(stub_pid_, &status, 0);
-  }
-
-  // First destroy the IPCControl. This makes sure that IPCControl is
-  // destroyed before the shared memory manager goes out of scope.
-  ipc_control_.reset();
-  stub_message_queue_.reset();
-  parent_message_queue_.reset();
-  memory_manager_.reset();
+  Stub()->TerminateStub(&decoupled_monitor_, &futures_, &thread_pool_);
```

Comment:
Instead of passing the arguments here, can we make sure that decoupled_monitor, futures, and thread_pool have exited properly here:

```
ModelState* model_state = reinterpret_cast<ModelState*>(Model());
      if (model_state->IsDecoupled()) {
        futures_.clear();
        Stub()-ParentMessageQueue()->Push(DUMMY_MESSAGE);
        decoupled_monitor_.join();
      }

      // Wait for all the futures to be finished.
      thread_pool_->wait();
```